src/MechaSwap.sol

```diff
@@ -0,0 +1,141 @@
+    /// @param amountX the amount of token X that'll be added to the pool
+    /// @return z the amount of LP tokens that'll be minted to the user
+    function addLiquidity(uint256 amountX) external returns(uint256 z){
+        require(x_0 > 0, "MechaSwap:Pool not intiated");
+        uint256 amountY = amountX * y_0 / x_0;
+        x_0 += amountX;
+        y_0 += amountY;
+        z = amountX / x_0;
+        tokenX.safeTransferFrom(msg.sender, address(this), amountX);
+        tokenY.safeTransferFrom(msg.sender, address(this), amountY);
+        _mint(msg.sender, z);
+        emit LiquidityProvided(msg.sender, amountX, amountY, z);
+    }
+
+    /// @notice used to remove tokens from the liquidity pool
+    /// @param amountZ the amount of LP tokens to burn
+    /// @return xOut the amount of token X returned to the user 
+    /// @return yOut the amount of token Y returned to the user
+    function removeLiquidity(uint256 amountZ) external returns(uint256 xOut, uint256 yOut) {
+        uint256 zOwnership = amountZ / _totalSupply;
+        xOut = WMul.wmul(x_0, zOwnership);
+        yOut = WMul.wmul(y_0, zOwnership);
+        _burn(msg.sender, amountZ);
+        tokenX.safeTransfer(msg.sender, xOut);
+        tokenY.safeTransfer(msg.sender, yOut);
+        emit LiquidityRemoved(msg.sender, xOut, yOut, amountZ);
+    }
+
+    /// @notice used to swap an amount of token X for token Y
+    /// @param xIn the amount of token X being sold to the pool
+    /// @return yOut the amount of token Y being bought from the pool
+    function swapXForY(uint256 xIn) external returns(uint256 yOut){
+        yOut = _price(xIn, x_0, y_0);
+        x_0 += xIn;
+        y_0 -= yOut;
+        tokenX.safeTransferFrom(msg.sender, address(this), xIn);
+        tokenY.safeTransfer(msg.sender, yOut);
+        emit Swap(msg.sender, xIn, 0, 0, yOut);
+    }
+
+    /// @notice used to swap amount of token Y for token X
+    /// @param yIn the amount of token Y being sold to the pool
+    /// @param xOut the amount of token X being bought from the pool
+    function swapYForX(uint256 yIn) external returns(uint256 xOut){
+        xOut = _price(yIn, y_0, x_0);
+        y_0 += yIn;
+        x_0 -= xOut;
+        tokenY.safeTransferFrom(msg.sender, address(this), yIn);
+        tokenX.safeTransfer(msg.sender, xOut);
+        emit Swap(msg.sender, 0, yIn, xOut, 0);
+    }
+
+    /// @notice internal function used for pricing token swaps
+    /// @param aIn amount of token that is being sold to the pool
+    /// @param aReserves reserves for the token that is being sold to the pool
+    /// @param bReserves rerserves for the token that is being bought from the pool
+    /// @param bOut amount of token that is being bought from the pool
+    function _price(uint256 aIn, uint256 aReserves, uint256 bReserves) internal view returns(uint256 bOut) {
+        require(x_0 > 0, "MechaSwap:Pool not intiated");
+        require(aIn > 0, "MechaSwap: Insufficient Tokens In");
```

Comment:
Same as above.

It is sometimes recommended to stop users from executing transactions without effect. While I'm not convinced this is necessary (or even advisable), such a check would be a `require` on the parameters of the external function.